File: crates/model/src/nav.rs
//! The equity curve and what it says: yearly time-weighted returns, the
//! annualized rate, and the drawdown.
//!
//! Returns are net of deposits and withdrawals throughout: money moved into or
//! out of the account is never counted as a gain or a loss.

use serde_json::{json, Map, Value};
use std::collections::BTreeMap;

use crate::dates::{days_between, shift_date};
use crate::value::{field_s, get, num, EPS};

#[derive(Clone, Debug)]
pub struct Point {
    pub d: String,
    pub v: f64,
    pub dep: Option<f64>,
}

impl Point {
    pub fn to_json(&self) -> Value {
        json!({"d": self.d, "v": self.v, "dep": self.dep})
    }
}

/// A number that is absent rather than zero, as Python's `_num(x, None)`.
fn opt_num(v: Option<&Value>) -> Option<f64> {
    match v {
        None | Some(Value::Null) => None,
        Some(x) => {
            let n = num(Some(x), f64::NAN);
            if n.is_nan() { None } else { Some(n) }
        }
    }
}

/// `model.equity_series`.
pub fn equity_series(points: &[Value]) -> Vec<Point> {
    let mut out: Vec<Point> = Vec::new();
    for p in points {
        let d: String = field_s(p, "date").chars().take(10).collect();
        let v = match opt_num(get(p, "equity")) { Some(v) => v, None => continue };
        if d.is_empty() {
            continue;
        }
        out.push(Point { d, v, dep: opt_num(get(p, "netDeposits")) });
    }
    out.sort_by(|a, b| a.d.cmp(&b.d));
    out
}

/// `model._nav_on`: the equity as of the end of a day.
fn nav_on(series: &[Point], day: &str) -> Option<f64> {
    let mut v = None;
    for p in series {
        if p.d.as_str() > day {
            break;
        }
        v = Some(p.v);
    }
    v
}

/// `model._deposits_on`: the running net deposits as of the end of a day.
fn deposits_on(series: &[Point], day: &str) -> Option<f64> {
    let mut v = None;
    for p in series {
        if p.d.as_str() > day {
            break;
        }
        if p.dep.is_some() {
            v = p.dep;
        }
    }
    v
}

pub struct YearSpan {
    pub r: f64,
    pub from: String,
    pub to: String,
    pub days: i64,
}

/// `model.year_return`: the daily chain-linked return for one calendar year,
/// net of deposits.
///
/// A balance under 1% of the account's peak is pre-history -- a few dollars
/// parked before the real start -- and a chain that began there would turn the
/// first real deposit into a wild return. The chain starts at the first point
/// clearing that floor, and the year is measured from there.
pub fn year_return(series: &[Point], year: &str, today: &str) -> Option<YearSpan> {
    if series.is_empty() {
        return None;
    }
    let cal = format!("{}-01-01", year);
    let year_end = format!("{}-12-31", year);
    let to = if year_end.as_str() < today { year_end } else { today.to_string() };

    let floor = series.iter().map(|p| p.v).fold(f64::NEG_INFINITY, f64::max) * 0.01;
    let start_day = shift_date(&cal, -1);
    let mut start = nav_on(series, &start_day);
    let mut after = start_day.clone();

    if !start.map_or(false, |s| s > floor) {
        let first = series.iter().find(|p| cal <= p.d && p.d <= to && p.v > floor)?;
        start = Some(first.v);
        after = first.d.clone();
    }
    let start = start?;

    let pts: Vec<&Point> = series.iter().filter(|p| p.d > after && p.d <= to).collect();
    if pts.is_empty() {
        return None;
    }

    let mut prev_eq = start;
    let mut prev_dep = deposits_on(series, &after);
    let mut factor = 1.0_f64;
    for p in pts {
        let eq = p.v;
        if !(prev_eq > 0.0) {
            return None;
        }
        let mut cf = 0.0;
        if let (Some(pd), Some(prev)) = (p.dep, prev_dep) {
            cf = pd - prev;
        }
        factor *= 1.0 + (eq - prev_eq - cf) / prev_eq;
        prev_eq = eq;
        if p.dep.is_some() {
            prev_dep = p.dep;
        }
    }
    let r = factor - 1.0;
    if r.is_nan() || r.is_infinite() {
        return None;
    }
    let span_from = if after == start_day { cal } else { after };
    let days = days_between(&span_from, &to);
    Some(YearSpan { r, from: span_from, to, days })
}

/// `model.benchmark_return`: the index over the same span the account's year
/// covers -- the calendar year, or from `start` when the account was funded
/// part way through it.
pub fn benchmark_return(bench: &BTreeMap<String, f64>, year: &str, today: &str, start: Option<&str>) -> Option<f64> {
    if bench.is_empty() {
        return None;
    }
    let cal = match start {
        Some(s) if !s.is_empty() => s.chars().take(10).collect::<String>(),
        _ => format!("{}-01-01", year),
    };
    let year_end = format!("{}-12-31", year);
    let to = if year_end.as_str() < today { year_end } else { today.to_string() };

    let mut prev: Option<f64> = None;
    let mut end: Option<f64> = None;
    for (d, v) in bench {
        if *d < cal {
            prev = Some(*v);
        } else if *d <= to {
            end = Some(*v);
        }
    }
    if prev.is_none() {
        let first = bench.iter().find(|(d, _)| **d >= cal && **d <= to)?;
        prev = Some(*first.1);
    }
    let prev = prev?;
    let end = end?;
    if prev == 0.0 {
        return None;
    }
    Some(end / prev - 1.0)
}

/// `model.yearly_returns`.
pub fn yearly_returns(series: &[Point], bench: &BTreeMap<String, f64>, today: &str) -> Vec<Value> {
    if series.is_empty() {
        return vec![];
    }
    let mut years: Vec<String> = series.iter().map(|p| p.d.chars().take(4).collect()).collect();
    years.sort();
    years.dedup();

    let peak = series.iter().map(|p| p.v).fold(f64::NEG_INFINITY, f64::max);
    let mut out = Vec::new();
    for y in years {
        // A year in which the account never held more than 1% of its peak is
        // pre-history, not a year of trading.
        let year_peak = series
            .iter()
            .filter(|p| p.d.starts_with(&y))
            .map(|p| p.v)
            .fold(0.0_f64, f64::max);
        if peak > 0.0 && year_peak < peak * 0.01 {
            continue;
        }
        let yr = match year_return(series, &y, today) { Some(v) => v, None => continue };
        let jan1 = format!("{}-01-01", y);
        let start_dep = deposits_on(series, &shift_date(&jan1, -1));
        let end_dep = deposits_on(series, &yr.to);
        let flow = match (start_dep, end_dep) {
            (Some(s), Some(e)) => Some(e - s),
            _ => None,
        };
        let end_v = nav_on(series, &yr.to);
        let sp_start = if yr.from != jan1 { Some(yr.from.as_str()) } else { None };
        out.push(json!({
            "year": y,
            "r": yr.r,
            "days": yr.days,
            "from": yr.from,
            "to": yr.to,
            "flow": flow,
            "endV": end_v,
            "spR": benchmark_return(bench, &y, today, sp_start),
        }));
    }
    out
}

/// `model.annualized`: the chain of usable years turned into a yearly rate.
pub fn annualized(years: &[Value]) -> Value {
    let mut prod = 1.0_f64;
    let mut days = 0_i64;
    let mut used: Vec<String> = Vec::new();
    for y in years {
        let r = match opt_num(get(y, "r")) { Some(r) => r, None => continue };
        let d = num(get(y, "days"), 0.0) as i64;
        if r <= -1.0 || d < 30 {
            continue;
        }
        prod *= 1.0 + r;
        days += d;
        used.push(field_s(y, "year"));
    }
    if days == 0 {
        return json!({"rate": Value::Null, "years": 0.0, "count": 0, "first": "", "last": ""});
    }
    let yrs = days as f64 / 365.25;
    let rate = if yrs >= 1.0 / 12.0 { prod.powf(1.0 / yrs) - 1.0 } else { prod - 1.0 };
    json!({
        "rate": rate,
        "years": yrs,
        "count": used.len(),
        "first": used.first().cloned().unwrap_or_default(),
        "last": used.last().cloned().unwrap_or_default(),
    })
}

/// `model._paired_flows`: the net deposit change per day, moved one day later
/// when the equity series only reflects the money the day after the deposit
/// record does.
fn paired_flows(series: &[Point]) -> Vec<f64> {
    let n = series.len();
    let mut flows = vec![0.0; n];
    for i in 1..n {
        let (p, prev) = (&series[i], &series[i - 1]);
        let (pd, prevd) = match (p.dep, prev.dep) { (Some(a), Some(b)) => (a, b), _ => continue };
        let cf = pd - prevd;
        if cf.abs() < EPS {
            continue;
        }
        let change_today = p.v - prev.v;
        if i + 1 < n {
            let change_next = series[i + 1].v - p.v;
            if (change_today - cf).abs() > (change_next - cf).abs() && change_today.abs() < cf.abs() * 0.5 {
                flows[i + 1] += cf;
                continue;
            }
        }
        flows[i] += cf;
    }
    flows
}

/// `model.drawdown`: the deepest fall of the flow-adjusted equity index.
pub fn drawdown(series: &[Point]) -> Value {
    if series.is_empty() {
        return json!({"pct": Value::Null, "abs": Value::Null, "at": "", "peakAt": ""});
    }
    let peak_v = series.iter().map(|p| p.v).fold(f64::NEG_INFINITY, f64::max);
    let floor = peak_v * 0.01;
    let mut idx = 1.0_f64;
    let mut prev: Option<&Point> = None;
    let mut peak_idx = 0.0_f64;
    let mut peak_at = String::new();
    let mut peak_equity = 0.0_f64;
    let mut dd = 0.0_f64;
    let mut dd_abs = 0.0_f64;
    let mut dd_at = String::new();
    let mut dd_peak_at = String::new();
    let flows = paired_flows(series);

    for (i, p) in series.iter().enumerate() {
        if let Some(pv) = prev {
            if pv.v > floor && pv.v > 0.0 {
                idx *= 1.0 + (p.v - pv.v - flows[i]) / pv.v;
            }
        }
        prev = Some(p);
        if p.v < floor {
            continue;
        }
        if idx >= peak_idx {
            peak_idx = idx;
            peak_at = p.d.clone();
            peak_equity = p.v;
        }
        if peak_idx <= 0.0 {
            continue;
        }
        let drop = idx / peak_idx - 1.0;
        if drop < dd {
            dd = drop;
            dd_abs = drop * peak_equity;
            dd_at = p.d.clone();
            dd_peak_at = peak_at.clone();
        }
    }
    json!({"pct": dd, "abs": dd_abs, "at": dd_at, "peakAt": dd_peak_at})
}

/// The benchmark map as the model reads it: dates to levels, in date order.
pub fn bench_map(v: Option<&Value>) -> BTreeMap<String, f64> {
    let mut out = BTreeMap::new();
    if let Some(Value::Object(m)) = v {
        for (k, val) in m {
            if let Some(f) = opt_num(Some(val)) {
                out.insert(k.clone(), f);
            }
        }
    }
    out
}

/// The equity series as the page receives it.
pub fn series_json(series: &[Point]) -> Vec<Value> {
    series.iter().map(|p| p.to_json()).collect()
}

/// Per-account equity series, keyed by the normalized nickname.
pub fn by_account(v: Option<&Value>) -> Map<String, Value> {
    let mut out = Map::new();
    if let Some(Value::Object(m)) = v {
        for (nick, pts) in m {
            let arr: Vec<Value> = pts.as_array().cloned().unwrap_or_default();
            let s = equity_series(&arr);
            out.insert(crate::value::norm_account_name(nick), Value::Array(series_json(&s)));
        }
    }
    out
}

File: go/internal/model/nav.go
package model

import (
	"math"
	"sort"

	"github.com/ProfessorBagholder/Bagholder/internal/py"
	"github.com/ProfessorBagholder/Bagholder/internal/store"
)

type YearRow struct {
	Year string   `json:"year"`
	R    float64  `json:"r"`
	Days int      `json:"days"`
	From string   `json:"from"`
	To   string   `json:"to"`
	Flow *float64 `json:"flow"`
	EndV *float64 `json:"endV"`
	SpR  *float64 `json:"spR"`
}

type Annualized struct {
	Rate  *float64 `json:"rate"`
	Years float64  `json:"years"`
	Count int      `json:"count"`
	First string   `json:"first"`
	Last  string   `json:"last"`
}

type Drawdown struct {
	Pct    *float64 `json:"pct"`
	Abs    *float64 `json:"abs"`
	At     string   `json:"at"`
	PeakAt string   `json:"peakAt"`
}

func EquitySeries(points []store.NavPoint) []EquityPoint {
	out := []EquityPoint{}
	for _, p := range points {
		d := cut(p.Date, 10)
		if d == "" || math.IsNaN(p.Equity) {
			continue
		}
		out = append(out, EquityPoint{D: d, V: p.Equity, Dep: copyPtr(p.NetDeposits)})
	}
	sort.SliceStable(out, func(i, j int) bool { return out[i].D < out[j].D })
	return out
}

func navOn(series []EquityPoint, day string) *float64 {
	var v *float64
	for i := range series {
		if series[i].D > day {
			break
		}
		v = py.Ptr(series[i].V)
	}
	return v
}

func depositsOn(series []EquityPoint, day string) *float64 {
	var v *float64
	for i := range series {
		p := &series[i]
		if p.D > day {
			break
		}
		if p.Dep != nil {
			v = copyPtr(p.Dep)
		}
	}
	return v
}

type yearSpan struct {
	r        float64
	from, to string
	days     int
}

func maxV(series []EquityPoint) float64 {
	m := math.Inf(-1)
	for _, p := range series {
		if p.V > m {
			m = p.V
		}
	}
	return m
}

func minStr(a, b string) string {
	if b < a {
		return b
	}
	return a
}

func yearReturn(series []EquityPoint, year, today string) *yearSpan {
	cal := year + "-01-01"
	to := minStr(year+"-12-31", today)
	if len(series) == 0 {
		return nil
	}
	floor := maxV(series) * 0.01
	startDay := shiftDate(cal, -1)
	start := navOn(series, startDay)
	after := startDay
	if !(start != nil && *start > floor) {
		var first *EquityPoint
		for i := range series {
			p := &series[i]
			if cal <= p.D && p.D <= to && p.V > floor {
				first = p
				break
			}
		}
		if first == nil {
			return nil
		}
		start = py.Ptr(first.V)
		after = first.D
	}
	var pts []EquityPoint
	for _, p := range series {
		if after < p.D && p.D <= to {
			pts = append(pts, p)
		}
	}
	if len(pts) == 0 {
		return nil
	}
	prevEq := *start
	prevDep := depositsOn(series, after)
	factor := 1.0
	for _, p := range pts {
		eq := p.V
		if !(prevEq > 0) {
			return nil
		}
		cf := 0.0
		if p.Dep != nil && prevDep != nil {
			cf = *p.Dep - *prevDep
		}
		factor *= 1 + (eq-prevEq-cf)/prevEq
		prevEq = eq
		if p.Dep != nil {
			prevDep = copyPtr(p.Dep)
		}
	}
	r := factor - 1
	if math.IsNaN(r) || math.IsInf(r, 0) {
		return nil
	}
	spanFrom := after
	if after == startDay {
		spanFrom = cal
	}
	return &yearSpan{r: r, from: spanFrom, to: to, days: daysBetween(spanFrom, to)}
}

func benchmarkReturn(bench map[string]float64, year, today, start string) *float64 {
	if len(bench) == 0 {
		return nil
	}
	days := store.SortedKeys(bench)
	cal := cut(start, 10)
	if cal == "" {
		cal = year + "-01-01"
	}
	to := minStr(year+"-12-31", today)
	var prev, end *float64
	for _, d := range days {
		if d < cal {
			prev = py.Ptr(bench[d])
		} else if d <= to {
			end = py.Ptr(bench[d])
		}
	}
	if prev == nil {
		for _, d := range days {
			if cal <= d && d <= to {
				prev = py.Ptr(bench[d])
				break
			}
		}
		if prev == nil {
			return nil
		}
	}
	if *prev == 0 || end == nil {
		return nil
	}
	return py.Ptr(*end / *prev - 1)
}

func YearlyReturns(series []EquityPoint, bench map[string]float64, today string) []YearRow {
	out := []YearRow{}
	if len(series) == 0 {
		return out
	}
	seen := map[string]bool{}
	var years []string
	for _, p := range series {
		y := cut(p.D, 4)
		if !seen[y] {
			seen[y] = true
			years = append(years, y)
		}
	}
	sort.Strings(years)
	peak := maxV(series)
	for _, y := range years {
		yearPeak := 0.0
		first := true
		for _, p := range series {
			if cut(p.D, 4) == y {
				if first || p.V > yearPeak {
					yearPeak = p.V
					first = false
				}
			}
		}
		if peak > 0 && yearPeak < peak*0.01 {
			continue
		}
		yr := yearReturn(series, y, today)
		if yr == nil {
			continue
		}
		startDep := depositsOn(series, shiftDate(y+"-01-01", -1))
		endDep := depositsOn(series, yr.to)
		var flow *float64
		if startDep != nil && endDep != nil {
			flow = py.Ptr(*endDep - *startDep)
		}
		endV := navOn(series, yr.to)
		spStart := ""
		if yr.from != y+"-01-01" {
			spStart = yr.from
		}
		out = append(out, YearRow{Year: y, R: yr.r, Days: yr.days, From: yr.from, To: yr.to, Flow: flow, EndV: endV, SpR: benchmarkReturn(bench, y, today, spStart)})
	}
	return out
}

func AnnualizedOf(years []YearRow) Annualized {
	prod := 1.0
	days := 0
	var used []string
	for _, y := range years {
		if y.R <= -1 || y.Days < 30 {
			continue
		}
		prod *= 1 + y.R
		days += y.Days
		used = append(used, y.Year)
	}
	if days == 0 {
		return Annualized{Rate: nil, Years: 0, Count: 0, First: "", Last: ""}
	}
	yrs := float64(days) / 365.25
	var rate float64
	if yrs >= 1.0/12 {
		rate = math.Pow(prod, 1/yrs) - 1
	} else {
		rate = prod - 1
	}
	return Annualized{Rate: py.Ptr(rate), Years: yrs, Count: len(used), First: used[0], Last: used[len(used)-1]}
}

func pairedFlows(series []EquityPoint) []float64 {
	n := len(series)
	flows := make([]float64, n)
	for i := 1; i < n; i++ {
		p, prev := series[i], series[i-1]
		if p.Dep == nil || prev.Dep == nil {
			continue
		}
		cf := *p.Dep - *prev.Dep
		if math.Abs(cf) < EPS {
			continue
		}
		changeToday := p.V - prev.V
		if i+1 < n {
			changeNext := series[i+1].V - p.V
			if math.Abs(changeToday-cf) > math.Abs(changeNext-cf) && math.Abs(changeToday) < math.Abs(cf)*0.5 {
				flows[i+1] += cf
				continue
			}
		}
		flows[i] += cf
	}
	return flows
}

func DrawdownOf(series []EquityPoint) Drawdown {
	if len(series) == 0 {
		return Drawdown{}
	}
	peakV := maxV(series)
	floor := peakV * 0.01
	idx := 1.0
	var prev *EquityPoint
	peakIdx := 0.0
	peakAt := ""
	peakEquity := 0.0
	dd, ddAbs := 0.0, 0.0
	ddAt, ddPeakAt := "", ""
	flows := pairedFlows(series)
	for i := range series {
		p := &series[i]
		if prev != nil && prev.V > floor && prev.V > 0 {
			idx *= 1 + (p.V-prev.V-flows[i])/prev.V
		}
		prev = p
		if p.V < floor {
			continue
		}
		if idx >= peakIdx {
			peakIdx = idx
			peakAt = p.D
			peakEquity = p.V
		}
		if peakIdx <= 0 {
			continue
		}
		drop := idx/peakIdx - 1
		if drop < dd {
			dd = drop
			ddAbs = drop * peakEquity
			ddAt = p.D
			ddPeakAt = peakAt
		}
	}
	return Drawdown{Pct: py.Ptr(dd), Abs: py.Ptr(ddAbs), At: ddAt, PeakAt: ddPeakAt}
}
